### PR TITLE
feat: new accounts route to show most downvoted ('controversial') posts

### PR DESCRIPTION
**Pull request by julianlam**

### CODE REVIEW

File: src/controllers/accounts/helpers.js [hunk 1]
```diff
@@ -155,7 +155,8 @@ async function getCounts(userData, callerUID) {
 	const cids = await categories.getCidsByPrivilege('categories:cid', callerUID, 'topics:read');
 	const promises = {
 		posts: db.sortedSetsCardSum(cids.map(c => `cid:${c}:uid:${uid}:pids`)),
-		best: db.sortedSetsCardSum(cids.map(c => `cid:${c}:uid:${uid}:pids:votes`)),
+		best: Promise.all(cids.map(async c => db.sortedSetCount(`cid:${c}:uid:${uid}:pids:votes`, 1, '+inf'))),
+		controversial: Promise.all(cids.map(async c => db.sortedSetCount(`cid:${c}:uid:${uid}:pids:votes`, '-inf', -1))),
```
Comment (julianlam) [inline src/controllers/accounts/helpers.js:159]: In lieu of a `sortedSetsCountSum`.

File: src/controllers/accounts/helpers.js [hunk 2]
```diff
@@ -169,6 +170,8 @@ async function getCounts(userData, callerUID) {
 		promises.blocks = user.getUserField(userData.uid, 'blocksCount');
 	}
 	const counts = await utils.promiseParallel(promises);
+	counts.best = counts.best.reduce((sum, count) => sum + count);
+	counts.controversial = counts.controversial.reduce((sum, count) => sum + count);
```
Comment (julianlam) [inline src/controllers/accounts/helpers.js:174]: The above returns the counts in an array, so they are summed here.

File: src/controllers/accounts/posts.js [hunk 3]
```diff
@@ -55,6 +56,27 @@ const templateToData = {
 			const cids = await categories.getCidsByPrivilege('categories:cid', callerUid, 'topics:read');
 			return cids.map(c => `cid:${c}:uid:${userData.uid}:pids:votes`);
 		},
+		getTopics: async (sets, req, start, stop) => {
+			let pids = await db.getSortedSetRevRangeByScore(sets, start, stop, '+inf', '1');
+			pids = await privileges.posts.filter('topics:read', pids, req.uid);
+			const postObjs = await posts.getPostSummaryByPids(pids, req.uid, { stripTags: false });
+			return { posts: postObjs, nextStart: stop + 1 };
+		},
```
Comment (julianlam) [inline src/controllers/accounts/posts.js:64]: The default `posts.getPostSummariesFromSet` call does not allow for `min` and `max` args, so I execute the same logic as that method, but call the `byScore` variant instead.

Not sure if there is a more elegant way you'd prefer.

File: src/controllers/accounts/posts.js [hunk 4]
```diff
@@ -55,6 +56,27 @@ const templateToData = {
...
+		getSets: async function (callerUid, userData) {
+			const cids = await categories.getCidsByPrivilege('categories:cid', callerUid, 'topics:read');
+			return cids.map(c => `cid:${c}:uid:${userData.uid}:pids:votes`);
+		},
+		getTopics: async (sets, req, start, stop) => {
+			let pids = await db.getSortedSetRangeByScore(sets, start, stop, '-inf', '-1');
+			pids = await privileges.posts.filter('topics:read', pids, req.uid);
+			const postObjs = await posts.getPostSummaryByPids(pids, req.uid, { stripTags: false });
+			return { posts: postObjs, nextStart: stop + 1 };
+		},
```
Comment (julianlam) [inline src/controllers/accounts/posts.js:79]: Ditto.

File: src/posts/votes.js [hunk 5]
```diff
@@ -258,7 +258,7 @@ module.exports = function (Posts) {
 		const topicData = await topics.getTopicFields(postData.tid, ['mainPid', 'cid', 'pinned']);
 
 		if (postData.uid) {
-			if (postData.votes > 0) {
+			if (postData.votes !== 0) {
```
Comment (julianlam) [inline src/posts/votes.js:261]: Saving upvotes **and** downvotes to the `:votes` set. Will need an upgrade script to update the zset to include posts with negative vote counts.

**Review by julianlam — COMMENTED** [review]
Added some clarification comments for review @barisusakli

File: src/controllers/accounts/helpers.js (hunk 1 above)
Comment (barisusakli) [reply src/controllers/accounts/helpers.js:159]: This needs to sum all the counts returned, `db.sortedSetsCardSum` returns a single number

**Comment by julianlam** [comment]
* [ ] Upgrade script to add posts with negative votes to `*:votes` zset
* [ ] Tests

**Comment by julianlam** [comment]
@barisusakli two concerns I have are:

1. Whether `Promise.all(cids.map(async c => db.sortedSetCount(`cid:${c}:uid:${uid}:pids:votes`, 1, '+inf'))),` will scale across 100+ categories, or whether a `sortedSetCountSum` should be created
2. Whether my addition of the `getTopic` property was the right thing to do

I think in both cases we're ok, but wanted to run it by you first.

**Comment by barisusakli** [comment]
1. not worth it I think, it would be similar in mongodb, since we cant count multiple zsets in one query
2. I think we can make a new function `getPosts` and use that instead of using the `getTopics` one

Not sure why auth tests are failing though 😆

**Pull request merged**